ui/pages/confirmations/components/confirm/info/approve/edit-spending-cap-modal/edit-spending-cap-modal.tsx

```diff
@@ -0,0 +1,177 @@
+import { TransactionMeta } from '@metamask/transaction-controller';
+import { BigNumber } from 'bignumber.js';
+import React, { useEffect, useMemo, useState } from 'react';
+import { useDispatch } from 'react-redux';
+import { hexToDecimal } from '../../../../../../../../shared/modules/conversion.utils';
+import {
+  Modal,
+  ModalBody,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  Text,
+  TextField,
+  TextFieldType,
+} from '../../../../../../../components/component-library';
+import {
+  AlignItems,
+  Display,
+  FlexDirection,
+  JustifyContent,
+  TextColor,
+  TextVariant,
+} from '../../../../../../../helpers/constants/design-system';
+import { useI18nContext } from '../../../../../../../hooks/useI18nContext';
+import { updateEditableParams } from '../../../../../../../store/actions';
+import { getCustomTxParamsData } from '../../../../../confirm-approve/confirm-approve.util';
+import { useConfirmContext } from '../../../../../context/confirm';
+import { useAssetDetails } from '../../../../../hooks/useAssetDetails';
+import { useApproveTokenSimulation } from '../hooks/use-approve-token-simulation';
+import { useTransactionGasEstimate } from '../hooks/use-transaction-gas-estimate';
+
+export const getAccountBalance = (userBalance: string, decimals: string) =>
+  new BigNumber(userBalance)
+    .dividedBy(new BigNumber(10).pow(Number(decimals)))
+    .toNumber();
+
+export const EditSpendingCapModal = ({
+  isOpenEditSpendingCapModal,
+  setIsOpenEditSpendingCapModal,
+}: {
+  isOpenEditSpendingCapModal: boolean;
+  setIsOpenEditSpendingCapModal: (newValue: boolean) => void;
+}) => {
+  const t = useI18nContext();
+
+  const dispatch = useDispatch();
+
+  const { currentConfirmation: transactionMeta } = useConfirmContext() as {
+    currentConfirmation: TransactionMeta;
+  };
+
+  const { userBalance, tokenSymbol, decimals } = useAssetDetails(
+    transactionMeta.txParams.to,
+    transactionMeta.txParams.from,
+    transactionMeta.txParams.data,
+  );
+
+  const accountBalance = getAccountBalance(userBalance || '0', decimals || '0');
+
+  const { formattedSpendingCap } = useApproveTokenSimulation(
+    transactionMeta,
+    decimals || '0',
+  );
+
+  const [customSpendingCapInputValue, setCustomSpendingCapInputValue] =
+    useState('');
+  const [customSpendingCap, setCustomSpendingCap] = useState('');
```

Comment:
I'm confused by these two pieces of state.

The text box updates `customSpendingCapInputValue` but `customSpendingCap` is only set when submit is clicked?

If that's the case, would it not be simpler to only keep `customSpendingCapInputValue` as state.

And then do any validation and `dispatch` imperatively inside the `handleSubmit` rather than during render via a `useEffect`?

Reply:
> The text box updates customSpendingCapInputValue but customSpendingCap is only set when submit is clicked?

> If that's the case, would it not be simpler to only keep customSpendingCapInputValue as state.

`customSpendingCapInputValue` is the value on the text input inside the modal. It needs to be a positive number.

`customSpendingCap` is set when the save button on the modal is pressed. This is the spending cap shown in the confirmation.

<img width="376" alt="Screenshot 2024-09-10 at 10 29 09" src="https://github.com/user-attachments/assets/52c5526e-852a-40eb-a593-8aed2aa05004">

> And then do any validation and dispatch imperatively inside the handleSubmit rather than during render via a useEffect?

Indeed, now that I moved a lot more of the logic to the modal component, it makes sense to move the dispatch inside the handleSubmit directly 👍 

The validation needs to be tied to the `customSpendingCap` to avoid showing a negative number in the confirmation screen (although the user can type it in the modal input as there's no way to limit at that level currently).

In summary:

customSpendingCapInputValue (Modal Input) -> Press "Save" -> coerce to positive number -> customSpendingCap (Confirmation) -> Press "Confirm" -> dispatch

Reply:
We can still generate the required custom data via the hook based on the input and reference those values within the callback. But if the `dispatch` is fundamentally only done on submit, then would it not be more explicit and more readable to actually call `dispatch` from within `handleSubmit` rather than a `useEffect`?

My confusion is why `customSpendingCap` needs to be state at all, can it not be derived in a standard variable and then rendered?

Why does it need to be persisted or generated at all if we hide the modal on submit?

Reply:
> We can still generate the required custom data via the hook based on the input and reference those values within the callback. But if the dispatch is fundamentally only done on submit, then would it not be more explicit and more readable to actually call dispatch from within handleSubmit rather than a useEffect?

I tried this approach but ran into a limitation due to needing to wait for the async hook that calculates the gas limit.

I worked around it by polling inside handleSubmit until the gas limit was calculated, but that is fundamentally flawed because the user can then confirm the confirmation before the dispatch is sent, resulting in a failure or inconsistent state. Here's the code I tested:

```tsx
 const { estimatedGasLimit } = useTransactionGasEstimate(
    transactionMeta,
    customTxParamsData,
    customSpendingCapInputValue,
  );

  const [isGasEstimateReady, setIsGasEstimateReady] = useState(false);

  useEffect(() => {
    if (estimatedGasLimit) {
      setIsGasEstimateReady(true);
    } else {
      setIsGasEstimateReady(false);
    }
  }, [estimatedGasLimit]);

  const handleCancel = useCallback(() => {
    setIsOpenEditSpendingCapModal(false);
    setCustomSpendingCapInputValue('');
    setIsGasEstimateReady(false);
  }, [setIsOpenEditSpendingCapModal, setCustomSpendingCapInputValue]);

  const waitForCondition = async (
    condition: boolean,
    interval: number,
    callback: () => void,
  ): Promise<void> => {
    return new Promise((resolve, reject) => {
      const checkCondition = async () => {
        try {
          if (condition) {
            callback();
            resolve();
          } else {
            setTimeout(checkCondition, interval);
          }
        } catch (error) {
          reject(error);
        }
      };

      checkCondition();
    });
  };

  const handleSubmit = useCallback(async () => {
    await waitForCondition(!isGasEstimateReady, 150, () => {
      dispatch(
        updateEditableParams(transactionMeta.id, {
          data: customTxParamsData,
          gas: hexToDecimal(estimatedGasLimit as string),
        }),
      );

      setIsOpenEditSpendingCapModal(false);
      setCustomSpendingCapInputValue('');
    });
  }, [
    isGasEstimateReady,
    customTxParamsData,
    estimatedGasLimit,
    setIsOpenEditSpendingCapModal,
    setCustomSpendingCapInputValue,
  ]);
```

I still believe the best solution is using the useEffect.

> My other confusion is why customSpendingCap needs to be state at all, can it not be derived in a standard variable and then rendered?

customSpendingCap functions as an intermediate state between the input field (setCustomSpendingCapInputValue) and the decoded spending cap inside the transactionMeta. It is used for displaying the value on the confirmation, to coerce negative input values into a spending cap of "0". This intermediate value is also important to trigger the dispatch, and without it we would have to dispatch every time the user types on the input field, which is an overkill in this case.

> Why does it need to be persisted or generated at all if we hide the modal on submit?

As per the comment below (https://github.com/MetaMask/metamask-extension/pull/26845#discussion_r1751628982), when the modal hides the component is not unmounted.

Reply:
I added the dispatch and the calls to the util methods directly to the handleSubmit method as discussed. I also added a loading indicator to the save button.